design/open/graveler_stateful_entities.md

```diff
@@ -0,0 +1,69 @@
+# Graveler Stateful Entities
+
+## Abstract
+
+Transitioning from PostgresSQL based metadata management DB to [generic KV Store design](https://github.com/treeverse/lakeFS/blob/master/design/accepted/metadata_kv/index.md) 
+introduced challenges in regard to atomicity and error handling.
+Previously, we relied on the transactional property of PSQL, to perform all-or-nothing operations to perform flows which requires 
+the modification of multiple entities atomically. Together with the built-in lock mechanism we were able to perform atomic, and concurrently safe operations.
+In the KV design, we rely on a very narrow interface, which does not support transactions. Additionally, the KV design does not rely on locks for operations.
+This complicates atomicity, error handling and rollback operations.
+This design can also pave to way for future support of [long-running hooks](https://github.com/treeverse/lakeFS/issues/3020), which require entities to be in an initial in-accessible/operable state,
+until the pre-action hooks are completed, and in case of failure deleted without implications on the end user.
+Proceeding the [stateful design for repositories](https://github.com/treeverse/lakeFS/pull/3661) , this document suggests adding states for all (most) of the graveler entities to 
+support both goals.
+
+## Supported Entities
+
+All entities which currently support pre/post actions on creation and deletion must support states. These include:
+* Branches
+* Commits
+* Tags
+
+All entities which require support of transactional / concurrent / rollback operations:
+* Repositories
+* Branches
+
+## Supported States
+
+### `INIT`
+
+The entity was created, in DB but not accessible to the user. Will move to `ACTIVE` state until initialization is completed. 
+
+### `ACTIVE`
+
+The entity is in a valid state and accessible to the user
+
+### `FAILED` / `INACTIVE`
+
+Terminal state - Inaccessible to the user, and a candidate for deletion
+
+## Error Handling
+
+Since error handling is defined the specific use case it will not be discussed under the current document, but generally 
+we will have 2 types of error handling:
+1. Setting entity to `FAILED` state for future deletion
+2. Performing a rollback operation for all the actions 
+3. Retry mechanism
```

Comment:
3 types here :-)

IMO errors (everywhere) should be split into a _message part_ and a _disposition part_.

The message part is intended for operators and users.  So it is detailed, and might be split into "visible on the backend" and "returned to the user" bits.  Because of this it is useless for programs.

The disposition part is intended for the calling program.  It is extremely **undetailed**, and consists of multiple property fields that says what to do in various cases.  In fact, the only property that is really needed is "ephemeral": is there any point in retrying this operation if nothing else has changed?  So for instance "branch not found" is _not_ ephemeral: if nothing changes then there is no point in looking for the branch again.  However "staging token changed while writing object to branch" _is_ ephemeral: if you try (everything) again then you might succeed!

The reason for talking about "ephemeral error" rather than "retrying" (at the lowest level) is that retries are determined by the client as much as they are by the server.  So a client can add a retry policy, but retry only on ephemeral failures.  It can still have its retry policy controlling everything else :-)